homeassistant/components/mqtt/notify.py

```diff
@@ -0,0 +1,393 @@
+        device = device_registry.async_get_device(
+            {(DOMAIN, id_) for id_ in config[CONF_DEVICE][CONF_IDENTIFIERS]},
+            {tuple(x) for x in config[CONF_DEVICE][CONF_CONNECTIONS]},
+        )
+
+        device_id = device.id
+
+    service = MqttNotificationService(
+        hass,
+        config[CONF_COMMAND_TOPIC],
+        MqttCommandTemplate(config.get(CONF_COMMAND_TEMPLATE), hass=hass),
+        config[CONF_ENCODING],
+        config.get(CONF_NAME),
+        config[CONF_QOS],
+        config[CONF_RETAIN],
+        config[CONF_TARGETS],
+        config[CONF_TITLE],
+        discovery_hash=discovery_hash,
+        device_id=device_id,
+        config_entry=config_entry,
+    )
+    services[service_name] = service
+
+    await service.async_setup(hass, service_name, service_name)
+    await service.async_register_services()
+
+
+def has_notify_services(hass: HomeAssistant, device_id: str) -> bool:
+    """Check if the device has registered notify services."""
+    services = hass.data.setdefault(MQTT_NOTIFY_SERVICES_SETUP, {})
+    for key, service in services.items():  # pylint: disable=unused-variable
+        if service.device_id == device_id:
+            return True
+    return False
+
+
+async def async_get_service(
+    hass: HomeAssistant,
+    config: ConfigType,
+    discovery_info: DiscoveryInfoType | None = None,
+) -> MqttNotificationService | None:
+    """Prepare the MQTT notification service through configuration.yaml."""
+    name = config.get(CONF_NAME)
+    if CONF_NAME not in config:
+        config[CONF_NAME] = DOMAIN
+    service_name = slugify(name or DOMAIN)
+    has_services = hass.services.has_service(notify.DOMAIN, service_name)
+    services = hass.data.setdefault(MQTT_NOTIFY_SERVICES_SETUP, {})
+    if service_name in services.keys() or has_services:
+        _LOGGER.error(
+            "Notify service '%s' is not unique, cannot register service(s)",
+            service_name,
+        )
+        return None
+    await async_setup_reload_service(hass, DOMAIN, PLATFORMS)
+    services[service_name] = MqttNotificationService(
+        hass,
+        config[CONF_COMMAND_TOPIC],
+        MqttCommandTemplate(config.get(CONF_COMMAND_TEMPLATE), hass=hass),
+        config[CONF_ENCODING],
+        name,
+        config[CONF_QOS],
+        config[CONF_RETAIN],
+        config[CONF_TARGETS],
+        config[CONF_TITLE],
+    )
+    return services[service_name]
+
+
+class MqttNotificationServiceUpdater:
+    """Add support for auto discovery updates."""
+
+    def __init__(self, hass: HomeAssistant, service: MqttNotificationService) -> None:
+        """Initialize the update service."""
+
+        async def async_discovery_update(
+            discovery_payload: DiscoveryInfoType | None,
+        ) -> None:
+            """Handle discovery update."""
+            async_dispatcher_send(
+                hass, MQTT_DISCOVERY_DONE.format(service.discovery_hash), None
+            )
```

Comment:
This needs to be done last, the point of sending that signal is to make it clear it's OK for `async_discovery_update` to be called again. Without that synchronization, things breaks if we receive two discovery messages immediately after each other.

Reply:
Ok, I'll have a look if this can be done after removing the discovery hash and cleanup.

Reply:
Was moved to the end